netutils/webclient/webclient.c

```diff
@@ -910,131 +1040,184 @@ int webclient_perform(FAR struct webclient_context *ctx)
+                          /* We don't have data to give to the client yet. */
+                        }
+                      else if (ctx->sink_callback)
+                        {
+                          ret = ctx->sink_callback(&ws->buffer, ws->offset,
+                                                   ws->datend, &ws->buflen,
+                                                   ctx->sink_callback_arg);
+                          if (ret != 0)
+                            {
+                              goto errout_with_errno;
+                            }
+                        }
+                      else
+                        {
+                          ctx->callback(&ws->buffer, ws->offset, ws->datend,
+                                        &ws->buflen, ctx->sink_callback_arg);
                         }
                     }
                   else
                     {
-                      ctx->callback(&ws->buffer, ws->offset, ws->datend,
-                                    &ws->buflen, ctx->sink_callback_arg);
-                    }
-                }
-              else
-                {
-                  redirected = true;
-                  nredirect++;
-                  if (nredirect > CONFIG_WEBCLIENT_MAX_REDIRECT)
-                    {
-                      nerr("ERROR: too many redirects (%u)\n", nredirect);
-                      goto errout;
-                    }
+                      ws->nredirect++;
+                      if (ws->nredirect > CONFIG_WEBCLIENT_MAX_REDIRECT)
+                        {
+                          nerr("ERROR: too many redirects (%u)\n",
+                               ws->nredirect);
+                          ret = -ELOOP;
+                          goto errout_with_errno;
+                        }
 
-                  break;
+                      ws->state = WEBCLIENT_STATE_CLOSE;
+                      ws->redirected = 1;
+                      break;
+                    }
                 }
             }
         }
 
-      conn_close(ctx, &conn);
+      if (ws->state == WEBCLIENT_STATE_CLOSE)
+        {
+          conn_close(ctx, conn);
+          ws->need_conn_close = false;
+          if (ws->redirected)
+            {
+              ws->state = WEBCLIENT_STATE_SOCKET;
+            }
+          else
+            {
+              ws->state = WEBCLIENT_STATE_DONE;
+            }
+        }
     }
-  while (redirected);
+  while (ws->state != WEBCLIENT_STATE_DONE);
 
   free(ws);
   return OK;
 
-errout:
-  ret = -errno;
 errout_with_errno:
-  conn_close(ctx, &conn);
+  if ((ctx->flags & WEBCLIENT_FLAG_NON_BLOCKING) != 0 &&
+      (ret == -EAGAIN || ret == -EINPROGRESS || ret == -EALREADY))
+    {
+      if (ret == -EINPROGRESS || ret == -EALREADY)
+        {
+          conn->flags |= CONN_WANT_WRITE;
+        }
+
+      return -EAGAIN;
+    }
+
+  if (ws->need_conn_close)
+    {
+      conn_close(ctx, conn);
+    }
 
   free(ws);
```

Comment:
```suggestion
  free(ws);
  ctx->ws = NULL;
```
To avoid use-after-free errors, the freed pointer should be nullified.

Reply:
@yamt https://en.wikipedia.org/wiki/Dangling_pointer